src/node_report.cc

```diff
@@ -0,0 +1,1026 @@
+#include <Lm.h>
+#include <tchar.h>
+#include <psapi.h>
+#else
+#include <sys/resource.h>
+// Get the standard printf format macros for C99 stdint types.
+#ifndef __STDC_FORMAT_MACROS
+#define __STDC_FORMAT_MACROS
+#endif
+#include <inttypes.h>
+#include <cxxabi.h>
+#include <dlfcn.h>
+#if defined(__linux__) || defined(__sun)
+#include <link.h>
+#endif
+#ifdef _AIX
+#include <sys/ldr.h>  // ld_info structure
+#endif
+// Include execinfo.h for the native backtrace API. The API is
+// unavailable on AIX and on some Linux distributions, e.g. Alpine Linux.
+#if !defined(_AIX) && !(defined(__linux__) && !defined(__GLIBC__))
+#include <execinfo.h>
+#endif
+#include <sys/utsname.h>
+#endif
+
+#ifdef __APPLE__
+#include <mach-o/dyld.h>  // _dyld_get_image_name()
+#endif
+
+
+
+#ifndef _WIN32
+extern char** environ;
+#endif
+
+namespace nodereport {
+
+using node::Environment;
+using node::Mutex;
+using node::NativeSymbolDebuggingContext;
+using v8::Exception;
+using v8::Function;
+using v8::HeapSpaceStatistics;
+using v8::HeapStatistics;
+using v8::Isolate;
+using v8::Local;
+using v8::MaybeLocal;
+using v8::Message;
+using v8::Number;
+using v8::RegisterState;
+using v8::SampleInfo;
+using v8::StackFrame;
+using v8::StackTrace;
+using v8::String;
+using v8::V8;
+using v8::Value;
+
+Mutex environ_mutex;
+
+// Internal/static function declarations
+static void WriteNodeReport(Isolate* isolate,
+                            Environment* env, DumpEvent event,
+                            const char* message, const char* location,
+                            const std::string& filename, std::ostream &out,
+                            MaybeLocal<Value> error, TIME_TYPE* time);
+static void PrintVersionInformation(std::ostream& out);
+static void PrintJavaScriptStack(std::ostream& out, Isolate* isolate,
+                                 DumpEvent event, const char* location);
+static void PrintJavaScriptErrorStack(std::ostream& out, Isolate* isolate,
+                                      MaybeLocal<Value> error);
+static void PrintStackFromStackTrace(std::ostream& out, Isolate* isolate,
+                                     DumpEvent event);
+static void PrintStackFrame(std::ostream& out, Isolate* isolate,
+                            Local<StackFrame> frame, int index, void* pc);
+static void PrintNativeStack(std::ostream& out);
+#ifndef _WIN32
+static void PrintResourceUsage(std::ostream& out);
+#endif
+static void PrintGCStatistics(std::ostream& out, Isolate* isolate);
+static void PrintSystemInformation(std::ostream& out);
+static void PrintLoadedLibraries(std::ostream& out);
+static void PrintComponentVersions(std::ostream& out);
+static void PrintSectionHeader(std::ostream& out, const char* header);
+
+// Global variables
+static std::atomic_int seq = {0};  // sequence number for report filenames
+const char* v8_states[] = {"JS", "GC", "COMPILER", "OTHER", "EXTERNAL", "IDLE"};
+static bool report_active = false;  // recursion protection
```

Comment:
per previous comment, I can move it to a better place or cover it with a mutex, which ever is better.

I have never encountered a recursive scenario. I am thinking about: when a `getReport` call is being serviced you get a fatal error (native assert or something) that leads upto a second report generation.

If you think this is an advanced scenario that can be addressed if we encounter those personally, I am happy to remove the logic around `report_active` for the time.

Reply:
If we do cover this with a mutex, we probably need to cover all of report generation with it – I’m not saying that’s a bad thing, but we’d need to be somewhat sure that we don’t introduce deadlock scenarios.

Ideally, I think the report-generating code should be written defensively enough to not trigger errors by itself. If the scenario you’re describing is realistic, then making this thread-local seems like a good idea? Otherwise, yeah, we may want to leave this out for now…

Reply:
done, removed this for now. I am thinking the presence of a dedicated thread (old code) that is subjected for repeated signal reception was the root cause of the recursion check. Now with the signal handling fully streamlined with core and libuv, I think recursion is impossible. Makes sense to you?

Reply:
further study around this area confirmed the above theory - the dedicated thread is prone to repeated signal reception; and can lead to re-entrance. All the current APIs in report runs synchronously on the main thread, so recursion is impossible.